### ISSUE TITLE
Dependencies do not get cached when using file uri scheme for repository

### ISSUE THREAD

**Issue by stefanleh** [gradle/gradle#8589]
### Expected Behavior

When using maven repository with file uri scheme for url i expect the artifacts to be downloaded to the gradle cache.

    repositories {
	    maven {
			url 'file://server/share/repo'
		}
    }

### Current Behavior

The artifacts get only referenced and not cached.
This is visible for example when using Buildship and checking the 'Project and External Dependencies' folder in Eclipse IDE.

### Context

I have to use a directory containing the dependecies for offline use (sealed environment).
To be able to define the dependecies the same way as in development, where i can use Sonatype Nexus, i have to pretend that this directory is a maven repository.

### Steps to Reproduce (for bugs)

Use Buildship and check 'Project and External Dependencies' folder in Eclipse IDE or use 'gradle cleanEclipse eclipse' and check .classpath file.

### Your Environment

Gradle 5.2.1
Windows 10 Pro

**Comment by ljacomet** [comment]
This is by designed, as there would be little benefit in duplicating files on the local drive.

Is it causing any issue aside from the difference in path in the IDE?

**Comment by stefanleh** [comment]
As showed in the example its a network share. And i want to avoid blocking the files on the share when the server gets started and the JVM holds handles to the jar files. Also access might be significant slower. Maybe a switch for changing the default behavior would be sufficient.

**Comment by stefanleh** [comment]
How about something simple like:

    repositories {
        maven {
            url 'file://server/share/repo'
            chached = true
        }
    }

**Comment by ljacomet** [comment]
Sorry I had missed the shared drive part in your sample.
It does looks simple indeed. 😉 

However it will hide a very important fact: local modifications, which nothing prevents, would be ignored by Gradle, unless Gradle verified the checksum of the metadata and the archive each time. And so while it will please a category of users, that flag will confuse others.

The recommendation from Gradle is to move that repo from a shared drive to something that can serve the files over an http connection _and_ enforce that released versions do not change, as is the default in repository managers. Or find a way to have a local only copy of it, that is synced now and then.

**Comment by bigdaz** [comment]
Related: https://github.com/gradle/gradle/issues/4322

**Comment by stefanleh** [comment]
I agree about the fact that released versions arent protected from change with this solution.
Albeit thats not really needed in my situation i see that this is an essential point in general.
Will try to use a simple HTTP server then.

Tnx for looking into the issue.

**Comment by bigdaz** [comment]
I think it would be great if all repository access was cached. I think the benefits of consistency and speed would outweigh the downsides of more disk usage. If excessive disk usage was a problem we could avoid copying the original files, but still do things like parse and cache the metadata.

So adding an opt-in for caching would be a good start, since we wouldn't break existing users that rely on the current "always refresh" behaviour.  Instead of calling it "cached", I'd call it "remote", so that Gradle treats it like any other remote (HTTP) repository.

**Comment by stefanleh** [comment]
As there are like 2 opinions on the matter now, should i close the issue or leave it open for further discussion?

**Comment by grv87** [comment]
+1 for this. The same problem. Network drive over VPN is very-very slow.

Old discussion of this issue:
https://discuss.gradle.org/t/how-can-i-force-caching-of-dependencies-from-a-file-based-repo/6214/2
There are some links to workarounds, but can't see them right now - issues.gradle.org is down.

**Comment by pahamala** [comment]
I have worked around this by syncing (ant.sync) the remote directory to a local directory in a gradle task and configuring the local directory as the file uri repository. The offline build works ok after one online build (i.e. sync done at least once) . 

The sync task:
```
task prepareLocalRepo() {
  doLast {
    if(!project.getGradle().startParameter.isOffline()) {
      ant.sync(todir: "<local dir>") {
        fileset(dir: "<remote dir>")
      }
    }
  }
}
```
